base/inference.jl

```diff
@@ -3394,24 +3397,35 @@ function effect_free(e::ANY, src::CodeInfo, mod::Module, allow_volatile::Bool)
             return true
         end
         ea = e.args
-        if head === :call && !isa(e.args[1], SSAValue) && !isa(e.args[1], Slot)
+        if head === :call
             if is_known_call_p(e, is_pure_builtin, src, mod)
                 if !allow_volatile
                     if is_known_call(e, arrayref, src, mod) || is_known_call(e, arraylen, src, mod)
                         return false
                     elseif is_known_call(e, getfield, src, mod)
+                        length(ea) == 3 || return false
                         et = exprtype(e, src, mod)
                         if !isa(et,Const) && !(isType(et) && isleaftype(et))
                             # first argument must be immutable to ensure e is affect_free
                             a = ea[2]
                             typ = widenconst(exprtype(a, src, mod))
-                            if !isa(typ, DataType) || typ.mutable || typ.abstract
+                            if isconstType(typ)
+                                if Const(:uid) ⊑ exprtype(ea[3], src, mod)
```

Comment:
yes, it's blacklist of one bad case but doesn't actually handle the general case

although it's possible we should just be more nuanced here. the `uid` field is probably the only observable value that can change, and it only can change when loading pre-compiled code – so we really only need to be concerned about constant folding it during then. for all other purposes, it's effect-free.